pkg/river/parser/internal.go

```diff
@@ -0,0 +1,614 @@
+package parser
+
+import (
+	"fmt"
+
+	"github.com/grafana/agent/pkg/river/ast"
+	"github.com/grafana/agent/pkg/river/scanner"
+	"github.com/grafana/agent/pkg/river/token"
+)
+
+// parser implements the River parser.
+//
+// It is only safe for callers to use exported methods as entrypoints for
+// parsing.
+//
+// Each Parse* and parse* method will describe the EBNF grammar being used for
+// parsing that nonterminal. The EBNF grammar will be written as LL(1) and
+// should directly represent the code.
+//
+// The parser will continue on encountering errors to allow a more complete
+// list of errors to be returned to the user. The resulting AST should be
+// discarded if errors were encountered during parsing.
+type parser struct {
+	file     *token.File
+	errors   ErrorList
+	scanner  *scanner.Scanner
+	comments []ast.CommentGroup
+
+	pos token.Pos   // Current token position
+	tok token.Token // Current token
+	lit string      // Current token literal
+}
+
+// newParser creates a new parser which will parse the provided src.
+func newParser(filename string, src []byte) *parser {
+	file := token.NewFile(filename)
+
+	p := &parser{
+		file: file,
+	}
+
+	p.scanner = scanner.New(file, src, func(pos token.Pos, msg string) {
+		p.errors.Add(&Error{
+			Position: file.PositionFor(pos),
+			Message:  msg,
+		})
+	}, 0)
+
+	p.next()
+	return p
+}
+
+// next advances the parser to the next non-comment token.
+func (p *parser) next() {
+	p.next0()
+
+	for p.tok == token.COMMENT {
+		p.consumeCommentGroup()
+	}
+}
+
+// next0 advances the parser to the next token. next0 should not be used
+// directly by parse methods; call next instead.
+func (p *parser) next0() { p.pos, p.tok, p.lit = p.scanner.Scan() }
```

Comment:
What is the naming of name0 indicate? nextToken make more sense?

Reply:
Better naming suggestions are definitely welcome, I took the pattern from go/parser but I don't really like it either. 

Basically:

* `next`: Gets next non-comment token 
* `next0`: (internal to `next`'s call tree) Get next token of any kind

Reply:
Oh, gotcha. I dont love it but the comment is clear.